tests/test_handler.py

```diff
@@ -0,0 +1,81 @@
+"""
+    Copyright 2019 Inmanta
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+    Contact: code@inmanta.com
+"""
+import pytest
+
+from inmanta import resources
+from inmanta.agent.handler import CRUDHandler, HandlerContext, ResourcePurged
+from inmanta.resources import Id, PurgeableResource, resource
+from utils import no_error_in_logs
+
+
+@pytest.mark.parametrize(
+    "purged_desired,purged_actual,excn,create,delete",
+    [
+        (True, False, True, False, False),
+        (True, True, False, False, False),
+        (True, False, False, False, True),
+        (True, True, True, False, False),
+        (False, False, True, True, False),
+        (False, True, False, True, False),
+        (False, True, True, True, False),
+        (False, False, False, False, False),
+    ],
+)
+@pytest.mark.parametrize(
+    "updated", [True, False],
+)
+def test_CRUD_handler_purged_response(purged_desired, purged_actual, excn, create, delete, updated, caplog):
```

Comment:
Why do we have so much redundancy on the parameters?

* `excn` is equal to `purged_actual`
* `create` can be derived from `purged_desired` and `purged_actual`
* `delete` can be derived from `purged_desired` and `purged_actual`

Reply:
1. excn _should_ be equal to purged_actual but they have completely different code paths  (the bug found only happens with purged_actual, not with excn). This test case explicitly verifies all combinations behave consistently
2. true, I don't know what is more (or less) readable

Reply:
Ok. Maybe add a comment explaining the difference between `excn` and `purged_actual`.